### ISSUE TITLE
Proposal: Beyond property initialization to code block initialization

### ISSUE THREAD

**Issue by softwaremills** [dotnet/csharplang#1243]
I have found this to be a really useful function:

	public static T With<T>(this T control, Action<T> setup)
	{
	    setup(control);
	    return control;
	}

I generally use it when I want to do some initialization without assigning a variable, and that initialization doesn't come in the form of setting a property:

    var views = new[]
    {
        new View(context).With(v => v.SetBackgroundColor(color1)),
        new View(context).With(v => v.SetBackgroundColor(color2)),
        new ViewGroup(context).With(vg =>
        {
            vg.AddView(new View(context).With(v => v.SetBackgroundColor(color1)));
            vg.AddView(new View(context).With(v => v.SetBackgroundColor(color2)));
        },
    };

That does work, but it annoys me that it's setting up a bunch of closure objects and functions behind the scenes for something that should be pretty simple syntactic sugar. Of course, not being a language designer, I have no idea what the proper syntax should be for this sort of thing. I imagine something like this:

    var views = new[]
    {
        new View(context) with v { v.SetBackgroundColor(color1); },
        new View(context) with v { v.SetBackgroundColor(color2); },
        new ViewGroup(context) with vg
        {
            vg.AddView(new View(context) with v { v.SetBackgroundColor(color1); });
            vg.AddView(new View(context) with v { v.SetBackgroundColor(color2); });
        },
    };

I mean, I use my little helper `With()` in places, but I still cringe. And I find this pattern so useful all over the place, especially where you're building some complex data structure and it's not sufficient to use property initialization or collection initialization.

**Comment by eyalsk** [comment]
@softwaremills C# has the concept of object initializers, maybe this concept can be expanded.

So it would be something like this:
```
var views = new[]
{
    new View(context) { SetBackgroundColor(color1) },
    new View(context) { SetBackgroundColor(color2) },
    new ViewGroup(context) {
        AddView(new View(context) { SetBackgroundColor(color1) }),
        AddView(new View(context) { SetBackgroundColor(color2) })
    },
};
```

**Comment by softwaremills** [comment]
@eyalsk That might get most of the way there. I like having the named reference to be able to have the full capability of blocks, like being able to, say, pass the `ViewGroup` to the created subviews, but your syntax is indeed closer to existing C#. In the `With` calls in my existing codebase, 16 out of 27 cases would be covered by that syntax. Some cases wouldn't be covered because of `if`s or `foreach`es in the block. Some wouldn't be covered because the `With` call doesn't actually have to be a `new`; it works just fine with any expression.

For instance, here's an actual snippet from my codebase:

	new UpdateShiftsData {
		DeletedBids = bids.Map(b => b.BidId),
		EditedShifts = new[] {
			new ShiftData {
				ShiftId = shift.ShiftId,
				WorkerId = bid.WorkerId,
				UpForBid = false,
				BidDeadlineMins = null,
				ModifiedDate = now,
			}.With(sd => {
				if (bid.SubShiftStart.HasValue) sd.Start = bid.SubShiftStart.GetValueOrDefault().FromUnixTime();
				if (bid.SubShiftFinish.HasValue) sd.Finish = bid.SubShiftFinish.GetValueOrDefault().FromUnixTime();
			}),
		},
	}.With(usd => {
		if (recalculateBreaks) {
			usd.DeletedBreaks = breaks.Map(b => b.BreakId);
			usd.AddedBreaks = applyBidBreaks.Map(b => new BreakData(b));
		}
	}).ToJObject()

**Comment by ufcpp** [comment]
with-expressions in https://github.com/dotnet/csharplang/blob/658e65403a9e14d9f614994d1fd6757d4fab9608/proposals/records.md

**Comment by softwaremills** [comment]
@ufcpp That is a good point; any proposed changes would need to be compatible with the syntax in records and, in particular, the with-expressions you mention. Good call.

This is quite different than records, of course, which are meant to enable good practices in immutable structs. This proposal is meant to refer to taking any expression, providing a block that operates on that expression, with the whole thing evaluating to the original expression. The syntax would ideally be general-purpose syntax that would be very useful for building complex literals. Such things happen a lot in the creation of data structures to be serialized (as in my code snippet above) or, possibly, applied to user interfaces (say, for instance, the creation of React-like virtual stand-in elements).

**Comment by reinspi** [comment]
@softwaremills hi!

Maybe it will be useful for you too
https://github.com/dotnet/csharplang/issues/803#issuecomment-361726730 with online samples https://dotnetfiddle.net/1sbITc

```c#
public static T With<T>(this T o, params object[] pattern) => o;
public static T To<T>(this T o, out T x) => x = o;
```
Some cases
```c#
	static Point Sample0() => new Point().To(out var p).With(
		p.X = 123,
		p.Y = 321,
		p.Name = "abc"
	);

	public static Point GetPoint() => new Point { Name = "Point Name" };
	static string NameProperty { get; set; }
	static string NameField;
	
	static void Sample1()
	{
		string nameLocal;
		GetPoint().To(out var p).With(
			p.X = 123,
			p.Y = 321,
			p.Name.To(out var name), /* right side assignment to the new variable */
			p.Name.To(out nameLocal), /* right side assignment to the declared var */
			NameField = p.Name, /* left side assignment to the declared variable */
			NameProperty = p.Name /* left side assignment to the property */
		);
		
		Console.WriteLine(name);
		Console.WriteLine(nameLocal);
		Console.WriteLine(NameField);
		Console.WriteLine(NameProperty);
	}

	static void Sample3() /* recursion */
	{
		GetPerson().To(out var p).With(
			p.Name.To(out var name),
			p.Subperson.To(out var p0).With(
				p0.Name.To(out var subpersonName0)
			),
			p.GetSubperson().To(out var p1).With( /* method return */
				p1.Name.To(out var subpersonName1)
			)
		);
		
		Console.WriteLine(subpersonName0);
		Console.WriteLine(subpersonName1);
	}
```
Bonus
- recursive in/out variables

Another interesting variations for your code
```c#
var views = new[]
{
    new View(context).To(out var v0).With(v0.Background = color0, v0.Foreground = color1),
    new View(context).To(out var v1).With(v1.Background = color1, v1.Foreground = color0),
    GetViewCollection().Merge(v0, v1) // new ViewCollection().Merge(v0, v1)
};
```
and
```c#
public static T[] Array<T>(params T[] items) => items;

var views = New.Array<object>(
    new View(context).To(out var v0).With(v0.Background = color0, v0.Foreground = color1),
    new View(context).To(out var v1).With(v1.Background = color1, v1.Foreground = color0),
    GetViewCollection().Merge(v0, v1) // new ViewCollection().Merge(v0, v1)
);
```
Enjoy if you like it!

**Comment by yaakov-h** [comment]
@reinspi Using slightly different extension methods doesn't solve the problem that this proposal is attempting to address:

> That does work, but it annoys me that it's setting up a bunch of closure objects and functions behind the scenes for something that should be pretty simple syntactic sugar.

**Comment by reinspi** [comment]
@yaakov-h 
The new suggested approach doesn't use closure objects and functions. All variables are opened at the top level, it is the key difference. Of course, nice to have the following syntax sugar too, but the range of related proposals has been declined.
```c#
		GetPoint() to var p? with {
			X = 123,
			Y = 321,
			Name to var name
		};
		
		Console.WriteLine(name);
		Draw(p);
```

**Comment by DavidArno** [comment]
@reinspi,

Hi ~~Makeloft~~ reinspi,

Please note, [your `To` method doesn't work with structs](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAIAYAEqCMBuAWAChjcA2bAJkwFEAPYYgb2MzewGZscKAVTAOoBLYAAsAPLwB8ACjFCAzpn4B7GJgAOAQzBaAtkpUAjAFYBTAMbAA2gF1NW4MDNgoASkwBeKZhWEi7JzcfMoqkrLySqrqKhDAyph0Ht6JXr54gcQAvsTECsCQVpgACipCUIxELAHsqFxwsUYANmaYABr+gXWYDRDNrQCanbVcuFgAcvpm/jkkRKjUAMLMrLU8JWUVmADK+hot6DLJPlBmAO4b5cBHAHS8KjKx8QBuOppuN8JiMsQAkL8aG5tNI4SgcGB/AE3AZpDiUHAQoj/QGTPStTyYABEWiMFkxfzc/lWbHI2AALNwABxHYmYaqBQKvMCaNK7PT7MyHQm0ro4ACcMgAJJi2p4mIC2ll1AMxYCBlLMKizLKbkqspjuTU2LNZoEgA). Did you forget to unit test again?

**Comment by yaakov-h** [comment]
@reinspi 

> The new suggested approach doesn't use closure objects and functions.

It still allocates unnecessary arrays.

**Comment by reinspi** [comment]
@DavidArno yes, thank, good note. But the simple following extension [fix this problem](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAIAYAEqCMBuAWAChjcA2bAJkwFEAPYYgb2MzewGZscKAVTAOoBLYAAsAPLwB8ACjFCAzpn4B7GJgAOAQzBaAtkpUAjAFYBTAMbAA2gF1NW4MDNgoASkwBeKZhWEi7JzcfMoqkrLySqrqKhDAyph0Ht6JXr54gazsWWyoXOTKAEqYADJmwJLqvIURoooJakWYAJ7JPs3+AL7ExArAkFaYAAoqQlCMRCwB7HmYcLFGADZmmAAa/oGz8xBLKwCaGzP5OFgAcvpmXT1EqNQAwsw5wcOj45gAyvoay+gybZhQMwAdxeY2AfwAdLwVDJYvEAG46TRuCHCMQyYgASExGghqzSOEoHBgWJxEL2aQ4lBwJKI2Nx5z0K08mAARFojBZWViUWVwRo3P4ngVUAAWbgADj+TymgUCiLAmjSnz03zMv0FT02OAAnDIACSs1aeJi41addR7E24vYWzCMszWiEOzqszXTNjdIhewJAA==)
```c#
public static TR Let<T, TR>(this T o, TR y) => y;
```

```c#
struct Point
{
    public double X;
    public double Y;
    public string Name;
}

    static Point Sample0() => new Point().To(out var p).With(
		p.X = 123,
		p.Y = 321,
		p.Name = "abc"
	).Let(p);
```

@yaakov-h 

> It still allocates unnecessary arrays.

By this reason had been suggested the syntax sugar optimized by compiler. Currently it is the bypass way for extending compiler limitations.

**Comment by terrajobst** [comment]
@reinspi 

Your comments look suspiciously similar to the ones posted by Makeloft, which we banned because he was spinning conversations out of control and demonstrated complete unwillingness to respond to any feedback about keeping threads on track. Thus, I've banned your account as well. If you believe this is as an error, please contact [Code of Conduct Reports](mailto:conduct@dotnetfoundation.org).

**Comment by reinspi** [comment]
@terrajobst 
Sorry, I believe that decision about Makeloft blocking was an error.

User shared with others valid information which may useful for someone.

**Comment by terrajobst** [comment]
@reinspi 

> User shared with others valid information which may useful for someone.

Please stop posting comments. I've blocked your new user account as well. It's clear based on community feedback that your comments are bordering on spam and significantly hinder our ability to engage with the community.

**Comment by reinspi** [comment]
@tannergooding 
Guys, I love you too. 

Sorry for this offtopic but it is possible to recreate a blocked account again and again.

Really I have friendly intentions.

Objectively is it my previous comments from this thread "are bordering on spam and significantly hinder our ability to engage with the community" ?

I just show the another workable way for `with` operator implementation/simulation. If you dislike it, don't use. Maybe it will be helpful for someone another.

If you don't want to see different accounts, please, unlock Makeloft. Most of my "spam" comments are just answers to another users. The new posts I write very rare only when I have to say something.

Hope for your understanding. Sincerely.

**Comment by masonwheeler** [comment]
@reinspi I don't think you understand.  If you've been blocked, it's because the community has decided they don't want you around.  Coming back under a new name turns you into a nuisance.  It will do nothing to improve things and everything to make people hate you.  It makes things worse for you, not better.

Go away and don't come back.  You've already lost; admit this with good grace, or you're just going to create problems for yourself and make people less and less willing to listen to your ideas, even if they have merit, because they're associated with someone who they find loathsome.

**Comment by reinspi** [comment]
@masonwheeler 
Seem that this repo is intended firstly for ideas and arguments and then for emotions.
But currently emotions are prevail. You may hate me, but your comments and attempts to block will be objective if you describe your opinion on suggested `with` operator simulation, not about me.

**Comment by masonwheeler** [comment]
Yes, it is intended firstly for ideas.  But your bad behavior crossed too many lines, and you were thrown out.  Sneaking back in is literally the worst thing you can do at this point, and I don't think you understand just how serious what you're doing is.

That's universally considered over the line, in all but the worst, most troll-infested communities.  Under some circumstances, people doing what you do have even had criminal charges filed against them.  You have no place here, and so you don't get to discuss your arguments or their merits anymore.  You've already lost that right.  Period.

**Comment by TAGC** [comment]
What happens when an unstoppable force meets an unbannable Makeloft?

**Comment by jnm2** [comment]
What happens? Everyone gets sick of the unsolicited advertising and vows never to use an extension method called `To`.

**Comment by silver-sd** [comment]
```csharp
```

A shorter syntax: 
`expr {{ .Member1 = someVar; }}`
The block follows the expression and is marked by special brackets for better readability and to remove the ambiguity. 
The access to the anonymous object members is possible by the dot prefix (to remove the ambiguity): `.Member`

Usages : 

1) Constructing and initializing the object tree. 
```csharp
var outerVar = 3;  
var lst = new List<C>()
{
    new C{InitOnlyProperty1 = 1},
    new C{InitOnlyProperty1 = 2},
    new C()
    {{      
        if(outerVar < 10) {} //can have any code as in the normal block
        .InitOnlyProperty1 = outerVar; // If the expression before the block is the constructor (without initializer) it is allowed to modify "initonly"
    }},
    new C{InitOnlyProperty1 = 4},

};
```
2) Modifying already created and initialized object tree, while preserving the structure of this tree for the code itself.
```csharp
var v1 = 1;
objLevel1.objLevel2.objLevel3
{{
    .Member1 = v1;
    .Member2 = v1;    
    for(i=0;i<v1;i++) {} //can have any code as in the normal block
    .objLevel4
    {{
        .Member1 = v1;
    	.Member2 = v1;
    	.objLevel5 
        {{ 
            .Member20 = 1;
            //...
        }};
        .obj2Level5 = new Cl(){{ .Member = 10;}};
    }};
    .Member20 = v1;
}};
```

**Comment by tomschrot** [comment]
Kotlin supports a similar construct, and has the automatic "it" variable in the callers' context for such closures.

I use such a extension for years, but naming it apply, config, utilize ....

The mechanism is also very useful as constructor and could replace the object initializer with the advantage to access the methods also.

`
var y = new XY (my =>
{
    my.value = 10;
    ...
});

class XY
{
    public XY () {}
    public XY (Action <XY> treat) => treat (this);

    public int value {get; set;} = 0;
}
`
